packages/connection/index.js

```diff
@@ -106,49 +107,50 @@ class Connection extends EventEmitter {
     return this.jid
   }
 
-  _online() {
-    this.emit('online', this.jid)
-  }
-
-  _authenticated() {
-    this.emit('authenticated')
+  _status(status, ...args) {
+    this.status = status
+    this.emit('status', status, ...args)
+    this.emit(status, ...args)
```

Comment:
So, we have two separate concerns here:
1. do something every time client gets 'online'  (i.e. light a green light)
2. do something when the connection process finishes (i.e. broadcast presence)

So both asynchronous event and promise are needed:
1. `client.on('online', jid => ....)` - called every time (possibly immediately if already online)
2. `client.when('online').then(jid => ....)` - resolved only once after online or immediately if already online

Does that make sense?